keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,696 @@
+In 3329, there was a decision to make kubelet transition pods to failed before deleting them.  This is feature toggled guarded by `PodDisruptionCondition`.  
+This means that when this feature is turned on, the job controller can count pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.
+If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.  
+To get around this problem we decided to add a field called terminating so we can count separately from active or failed.  
+
+Another issue is described here: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509.  
+If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.  Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new enum field:
+
+```golang
+// This field controls when we recreate pods
+// Default will be TerminatingOrFailed ie recreate pods when they are failed
+// +enum 
+type RecreatePodsWhen string
+const (
+ // This is a field that recreates pods when they are marked as terminating or failed
+ // ie this means that as soon as pods get marked for deletion they will be recreated
+ TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
+ // Only recreate pods when they are marked as failed.
+ Failed              RecreatePodsWhen = "Failed"
+)
+```
+
+```golang
+type JobSpec struct{
+  ...
+ // RecreatePodsWhen specifies when pods should be recreated.
+ // TerminatingOrFailed means to recreate when a pod is either terminating or failed
+ // TerminatingOrFailed is the default.
+ // Failed means to wait until pods are fully terminated or failed before recreating
+ // +optional
+ RecreatePodsWhen *RecreatePodsWhen
+}
+```
+
+So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.
+
+```golang
+type JobStatus struct {
+  ...
+  // Number of terminating pods
+  // +optional
+  terminating *int32
+}
+```
+
+We will allow only opt-in behavior for this feature so we will fall back to `TerminatingOrFailed` for `RecreatePodsWhen`.  
+
+### Implementation
+
+The Job Controller filters all pods and classifies the pods as active.  We will append terminating pods to list as part of this KEP.
+
+There are two cases in the job controller where deletions can happen programatically:
```

Comment:
TBH I don't understand why the distinction is made here (on pods being deleted by Job controller itself and those deleted by external controllers, such as scheduler during preemption or Kubelet on shutting down a node). 

Can we just say that we provide the new `.status.terminating` field that counts the number of terminating pods (with deletionTimestamp!= nil), but not terminal yet?

Reply:
I seem to remember someone asking how will this behave with `PodFailurePolicy` or `ActiveDeadlineSeconds` so I made a point to mention it.

Reply:
I see, makes sense, but then mention that "in particular, ...", to make it clear that the distinction isn't important from the design perspective. It is just to provide examples of scenarios covered. 

Also, IMO this does not belong to Design Details -> Implementation. I suggest this could go to the "Notes/Constraints/Caveats" section.

Reply:
Moved.